// ng-zorro-antd/esm2022/table/src/table/tfoot-summary.component.mjs
/**
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
import { NgTemplateOutlet } from '@angular/common';
import { booleanAttribute, ChangeDetectionStrategy, Component, inject, Input, ViewChild, ViewEncapsulation } from '@angular/core';
import { NzTableStyleService } from '../table-style.service';
import * as i0 from "@angular/core";
function fixedAttribute(value) {
    return value === 'top' || value === 'bottom' ? value : booleanAttribute(value) ? 'bottom' : null;
}
/* eslint-disable @angular-eslint/component-selector */
export class NzTfootSummaryComponent {
    constructor() {
        this.nzFixed = null;
        this.nzTableStyleService = inject(NzTableStyleService, { optional: true });
        this.isInsideTable = !!this.nzTableStyleService;
    }
    ngOnInit() {
        this.nzTableStyleService?.setTfootTemplate(this.templateRef);
    }
    ngOnChanges(changes) {
        const { nzFixed } = changes;
        this.nzTableStyleService?.setTfootFixed(nzFixed.currentValue);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.2", ngImport: i0, type: NzTfootSummaryComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.0.0", version: "18.1.2", type: NzTfootSummaryComponent, isStandalone: true, selector: "tfoot[nzSummary]", inputs: { nzFixed: ["nzFixed", "nzFixed", fixedAttribute] }, host: { properties: { "class.ant-table-summary": "!isInsideTable || !nzFixed" } }, viewQueries: [{ propertyName: "templateRef", first: true, predicate: ["contentTemplate"], descendants: true, static: true }], usesOnChanges: true, ngImport: i0, template: `
    <ng-template #contentTemplate>
      <ng-content></ng-content>
    </ng-template>
    @if (!isInsideTable || !nzFixed) {
      <ng-template [ngTemplateOutlet]="contentTemplate"></ng-template>
    }
  `, isInline: true, dependencies: [{ kind: "directive", type: NgTemplateOutlet, selector: "[ngTemplateOutlet]", inputs: ["ngTemplateOutletContext", "ngTemplateOutlet", "ngTemplateOutletInjector"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.2", ngImport: i0, type: NzTfootSummaryComponent, decorators: [{
            type: Component,
            args: [{
                    standalone: true,
                    selector: 'tfoot[nzSummary]',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                    template: `
    <ng-template #contentTemplate>
      <ng-content></ng-content>
    </ng-template>
    @if (!isInsideTable || !nzFixed) {
      <ng-template [ngTemplateOutlet]="contentTemplate"></ng-template>
    }
  `,
                    imports: [NgTemplateOutlet],
                    host: {
                        '[class.ant-table-summary]': '!isInsideTable || !nzFixed'
                    }
                }]
        }], propDecorators: { nzFixed: [{
                type: Input,
                args: [{ transform: fixedAttribute }]
            }], templateRef: [{
                type: ViewChild,
                args: ['contentTemplate', { static: true }]
            }] } });
