// ng-zorro-antd/esm2022/result/result-cells.mjs
/**
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
import { Directive } from '@angular/core';
import * as i0 from "@angular/core";
export class NzResultTitleDirective {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.2", ngImport: i0, type: NzResultTitleDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.1.2", type: NzResultTitleDirective, isStandalone: true, selector: "div[nz-result-title]", host: { classAttribute: "ant-result-title" }, exportAs: ["nzResultTitle"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.2", ngImport: i0, type: NzResultTitleDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'div[nz-result-title]',
                    exportAs: 'nzResultTitle',
                    host: {
                        class: 'ant-result-title'
                    },
                    standalone: true
                }]
        }] });
export class NzResultSubtitleDirective {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.2", ngImport: i0, type: NzResultSubtitleDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.1.2", type: NzResultSubtitleDirective, isStandalone: true, selector: "div[nz-result-subtitle]", host: { classAttribute: "ant-result-subtitle" }, exportAs: ["nzResultSubtitle"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.2", ngImport: i0, type: NzResultSubtitleDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'div[nz-result-subtitle]',
                    exportAs: 'nzResultSubtitle',
                    host: {
                        class: 'ant-result-subtitle'
                    },
                    standalone: true
                }]
        }] });
export class NzResultIconDirective {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.2", ngImport: i0, type: NzResultIconDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.1.2", type: NzResultIconDirective, isStandalone: true, selector: "i[nz-result-icon], div[nz-result-icon]", exportAs: ["nzResultIcon"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.2", ngImport: i0, type: NzResultIconDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'i[nz-result-icon], div[nz-result-icon]',
                    exportAs: 'nzResultIcon',
                    standalone: true
                }]
        }] });
export class NzResultContentDirective {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.2", ngImport: i0, type: NzResultContentDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.1.2", type: NzResultContentDirective, isStandalone: true, selector: "div[nz-result-content]", host: { classAttribute: "ant-result-content" }, exportAs: ["nzResultContent"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.2", ngImport: i0, type: NzResultContentDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'div[nz-result-content]',
                    exportAs: 'nzResultContent',
                    host: {
                        class: 'ant-result-content'
                    },
                    standalone: true
                }]
        }] });
export class NzResultExtraDirective {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.2", ngImport: i0, type: NzResultExtraDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.1.2", type: NzResultExtraDirective, isStandalone: true, selector: "div[nz-result-extra]", host: { classAttribute: "ant-result-extra" }, exportAs: ["nzResultExtra"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.2", ngImport: i0, type: NzResultExtraDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'div[nz-result-extra]',
                    exportAs: 'nzResultExtra',
                    host: {
                        class: 'ant-result-extra'
                    },
                    standalone: true
                }]
        }] });
